homeassistant/components/http.py

```diff
@@ -200,12 +201,22 @@ def _handle_request(self, method):  # pylint: disable=too-many-branches
                     "Error parsing JSON", HTTP_UNPROCESSABLE_ENTITY)
                 return
 
-        self.authenticated = (self.server.api_password is None or
-                              self.headers.get(HTTP_HEADER_HA_AUTH) ==
-                              self.server.api_password or
-                              data.get(DATA_API_PASSWORD) ==
-                              self.server.api_password or
-                              self.verify_session())
+        if self.verify_session():
+            # The user has a valid session already
+            self.authenticated = True
+        elif self.server.api_password is None:
+            # No password is set, so everyone is authenticated
+            self.authenticated = True
+        elif hmac.compare_digest(self.headers.get(HTTP_HEADER_HA_AUTH, ''),
+                                 self.server.api_password):
```

Comment:
While it's true that the purpose of compare_digest is ostensibly to compare digests, it doesn't actually care if it's a digest or not. It'll happy to constant time comparisons on any pair of strings. It just calls _tscmp. https://github.com/python/cpython/blob/2d264235f6e066611b412f7c2e1603866e0f7f1b/Modules/_operator.c#L176